### DISCUSSION TITLE
How can I add filters as a control button?

### DISCUSSION THREAD

**Discussion by uugaabuugaa** [tomasklaen/uosc#928] [Q&A] [answered]
trying to add `af toggle lavfi=[dynaudnorm=f=75:g=25:p=0.55:m=5:r=0.9:b=1]` as a control button but uosc completely stops working.  I'm assuming it has something to do with the extra `:`s being present in the command line. Is there a way to get it to work? Preferably not by using any lua scripts since this is the only filter I use based on whether it's a movie or music video.

**Comment by hooke007** [comment]
You can add the af in the profile first.
```
[profile1]
input-commands=af toggle lavfi=[dynaudnorm=f=75:g=25:p=0.55:m=5:r=0.9:b=1]
```
Then use the profile cmd instead in uosc conf
`apply-profile profile1`

**Reply by uugaabuugaa** [reply]
Using profile seems like an interesting solution. But the problem is I already have more than a few profiles based on youtube quality and scaling quality and have a button to cycle through them. Using the filter as a profile would mean I have to create two for each current profile with or without filter which doesn't seem like the right approach. I think using it as a conditional profile based on the length of the video on top of currently applied profile is better (although I don't know how). 

Also I tried
 ```
[profile1]
input-commands=af toggle lavfi=[dynaudnorm=f=75:g=25:p=0.55:m=5:r=0.9:b=1]
```
but doesn't seem to do anything. As a standalone button it doesn't show any osd of profile being applied. And adding to profile cycler button, it shows the profile as being applied but doesn't really apply the filter. I think if using a lua script to apply filter can be used as control button I'm willing to try it out.

**Reply by Sneakpeakcss** [reply, marked as answer]
I think this was the suggestion:

```
[profile1]
input-commands=af toggle lavfi=[dynaudnorm=f=75:g=25:p=0.55:m=5:r=0.9:b=1]
```

```
controls=command:graphic_eq:apply-profile profile1
```
___

There's an issue with this though. `mpv` doesn't seem to execute that command when the same profile is applied again, so it doesn't actually toggle on the second button press.

Doing something like this works:

```
[profile1]
input-commands=af toggle lavfi=[dynaudnorm=f=75:g=25:p=0.55:m=5:r=0.9:b=1]
input-commands=apply-profile emptyprofile
[emptyprofile]
```

But this feels like a unnecessary workaround for the issue where any command containing `:` is crashing `uosc`:

```
[uosc] 
[uosc] stack traceback:
[uosc] 	...pData\Roaming\mpv\scripts/uosc/elements/Controls.lua:27: in function 'init'
[uosc] 	...ers\user\AppData\Roaming\mpv\scripts/uosc/lib/std.lua:263: in function 'new'
[uosc] 	C:\Users\user\AppData\Roaming\mpv\scripts/uosc/main.lua:1150: in function '_commit'
[uosc] 	C:\Users\user\AppData\Roaming\mpv\scripts/uosc/main.lua:1142: in function <C:\Users\user\AppData\Roaming\mpv\scripts/uosc/main.lua:1139>
[uosc] 	C:\Users\user\AppData\Roaming\mpv\scripts/uosc/main.lua:1161: in main chunk
[uosc] 	[C]: at 0x7ffae16a62d0
[uosc] 	[C]: at 0x7ffae16a5130
[uosc] Lua error: ...pData\Roaming\mpv\scripts/uosc/elements/Controls.lua:120: attempt to call field 'error' (a nil value)
```
___
 
You can use a lua script if you want:

```lua
mp.register_script_message("toggle_dynaudnorm", function()
    mp.command("af toggle lavfi=[dynaudnorm=f=75:g=25:p=0.55:m=5:r=0.9:b=1]")
end)
```

```
command:graphic_eq:script-message toggle_dynaudnorm
```

**Reply by hooke007** [reply]
I think input-cmds is probably broken in somewhere.
Use af-toggle instead. It works here.
```
[profile1]
 af-toggle=lavfi=[dynaudnorm=f=75:g=25:p=0.55:m=5:r=0.9:b=1]
```

**Reply by uugaabuugaa** [reply]
> You can use a lua script if you want:
> 
> ```
> mp.register_script_message("toggle_dynaudnorm", function()
>     mp.command("af toggle lavfi=[dynaudnorm=f=75:g=25:p=0.55:m=5:r=0.9:b=1]")
> end)
> ```
> 
> ```
> command:graphic_eq:script-message toggle_dynaudnorm
> ```

The lua script works exactly as I want to. Thanks a lot! Not testing the profile method to not complicate my existing profiles further but appreciate the help. Changed the icon since uosc's audio button uses graphic_eq by default. Hopefully uosc devs will do something about the `:` in commands; maybe change something to ignore `:` while they're between `""` or something like that.
